### PR TITLE
Tweaks and improvements on SMIR around generics_of and predicates_of

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk 

This allows an API like the following ...

```rust
    let trait_decls = stable_mir::all_trait_decls().iter().map(|trait_def| {
        let trait_decl = stable_mir::trait_decl(trait_def);
        let generics = trait_decl.generics_of();
        let predicates = trait_decl.predicates_of().predicates;
```

I didn't like that much `trait_def.trait_decl()` which is it possible but adding a method to a def_id that loads up a whole trait definition looks backwards to me.

### CODE REVIEW

File: compiler/rustc_smir/src/stable_mir/mod.rs [hunk 1]
```diff
@@ -126,8 +126,8 @@ pub trait Context {
     fn trait_decl(&mut self, trait_def: &TraitDef) -> TraitDecl;
     fn all_trait_impls(&mut self) -> ImplTraitDecls;
     fn trait_impl(&mut self, trait_impl: &ImplDef) -> ImplTrait;
-    fn generics_of(&mut self, generic_def: &GenericDef) -> Generics;
-    fn predicates_of(&mut self, trait_def: &TraitDef) -> GenericPredicates;
+    fn generics_of(&mut self, def_id: rustc_span::def_id::DefId) -> Generics;
+    fn predicates_of(&mut self, def_id: rustc_span::def_id::DefId) -> GenericPredicates;
```
Comment (compiler-errors) [inline compiler/rustc_smir/src/stable_mir/mod.rs:130]: Shouldn't this be using the "stable" `DefId` defined on line 32 of this file? Seems surprising to be leaking `DefId` in the stable API here.

Comment (compiler-errors) [reply compiler/rustc_smir/src/stable_mir/mod.rs:130]: Or alternatively, take a `&CrateItem` like other "item-agnostic" fn above?

Comment (spastorino) [reply compiler/rustc_smir/src/stable_mir/mod.rs:130]: Yes, I was trying to show Oli the problem the current approach has. We need a trait on smir thst each different def implement or we need to reuse some existing type or just pass smir defid

**Comment by spastorino** [comment]
I'm leaving the PR this way for @oli-obk to check it out, but I'm going to add a macro so we can have this *Def new type wrappers defined over `DefId` implementing also `IntoDefId` automatically.

**Comment by spastorino** [comment]
@oli-obk maybe you don't want the last commit or maybe we want to keep implementing `Index` for the *Def wrappers.

**Pull request merged**